rectools/models/sasrec.py

```diff
@@ -1179,6 +1288,15 @@ def __init__(  # pylint: disable=too-many-arguments, too-many-locals
         self.loss = loss
         self.gbce_t = gbce_t
 
+    @property
+    def requires_negatives(self) -> bool:
+        """Indicator whether negatives are needed for loss computation during training."""
+        if isinstance(self.loss, SessionEncoderHeadBase):
+            return self.loss.requires_negatives
+        if self.loss == "softmax":
```

Comment:
You have this property in every class, why do you need to have this logic here again?

Reply:
we don't have SessionEncoderHead initialised from LossName yet. so we can't say if it requires negatives from initialised Head.
But we need to know this for data_preaparator initialisation. And data preparator is needed before constructing torch nets because torch nets depend on data_preaparator results.

Reply:
Sorry the code is too complicated for me to deep dive into it now. The only thing I know that having duplicated logic is a really bad idea. Let's please don't do this

Reply:
We really can't not do this.
We do not initialise torch nets until we get dataset (and Head is a torch net).
We can't get dataset until we get data_preaparator.
We can't get data_preparator until we know if need negatives or not.

Many things depend in whether we need negatives or not.

Reply:
We can initialise heads during model initialisation just to get `require_negatives` value. And save received loss value for `get_config`. But I'm not sure it's better this way.